tools/deprecated-config-checker/checker/checker.go

```diff
@@ -0,0 +1,289 @@
+}
+
+func appenToPath(path, key string) string {
+	if path == "" {
+		return key
+	}
+	return path + "." + key
+}
+
+func getLimitsConfig(in RawYaml) RawYaml {
+	limits, exists := in["limits_config"]
+	if !exists {
+		return nil
+	}
+	return limits.(RawYaml)
+}
+
+func getOverrides(runtimeConf RawYaml) RawYaml {
+	overrides, exists := runtimeConf["overrides"]
+	if !exists {
+		return nil
+	}
+	return overrides.(RawYaml)
+}
+
+func checkRuntimeConfigDeprecated(deprecates, runtimeConfig RawYaml) []DeprecationNotes {
+	deprecatedLimits := getLimitsConfig(deprecates)
+	if deprecatedLimits == nil {
+		return nil
+	}
+
+	overrides := getOverrides(runtimeConfig)
+	if overrides == nil {
+		return nil
+	}
+
+	// We check the deprecated fields for each tenant
+	var deprecations []DeprecationNotes
+	for tenant, tenantOverrides := range overrides {
+		tenantDeprecations := checkConfigDeprecated(deprecatedLimits, tenantOverrides.(RawYaml))
+		for i := range tenantDeprecations {
+			tenantPath := appenToPath("overrides", tenant)
+			tenantDeprecations[i].ItemPath = appenToPath(tenantPath, tenantDeprecations[i].ItemPath)
+		}
+		deprecations = append(deprecations, tenantDeprecations...)
+	}
+
+	return deprecations
+}
+
+func checkConfigDeprecated(deprecates, config RawYaml) []DeprecationNotes {
+	return enumerateDeprecatesFields(deprecates, config, "", []DeprecationNotes{})
+}
+
+func enumerateDeprecatesFields(deprecates, input RawYaml, rootPath string, deprecations []DeprecationNotes) []DeprecationNotes {
+	for key, deprecate := range deprecates {
+		inputValue, exists := input[key]
+		if !exists {
+			// If this item is not set in the config, we can skip it.
+			continue
+		}
+
+		path := appenToPath(rootPath, key)
+
+		note, isDeprecatedNote := getDeprecationAnnotation(deprecate)
+		if isDeprecatedNote {
+			var inputDeprecated bool
+
+			// If there are no specific values deprecated, the whole config is deprecated.
+			// Otherwise, look for the config value in the list of deprecated values.
+			if len(note.DeprecatedValues) == 0 {
+				inputDeprecated = true
+			} else {
+				for _, v := range note.DeprecatedValues {
+					if v == inputValue {
+						inputDeprecated = true
+						break
+					}
+				}
+			}
+
+			if inputDeprecated {
+				var itemValueStr string
+				if v, is := inputValue.(string); is {
+					itemValueStr = v
+				}
+
+				deprecations = append(deprecations, DeprecationNotes{
+					deprecationAnnotation: note,
+					ItemPath:              path,
+					ItemValue:             itemValueStr,
+				})
+				continue
+			}
+		}
+
+		// To this point, the deprecate item is not a leaf, so we need to recurse into it.
+		if deprecateYaml, is := deprecate.(RawYaml); is {
+			switch v := inputValue.(type) {
+			case RawYaml:
+				deprecations = enumerateDeprecatesFields(deprecateYaml, v, path, deprecations)
+			case []interface{}:
+				// If the config is a list, recurse into each item.
+				for i, item := range v {
+					itemYaml := item.(RawYaml)
```

Comment:
could this fail?

Reply:
This may panic if the item is a list of strings or numbers. Looking at the config, looks like we do have lists of strings (e.g. `distributor.ring.instance_interface_names`). 

But this will not happen because we iterate through the deprecated/deleted fields. So if the field is deprecated, we won't recurse into it.

We need to update how we handle the deprecation early on to be able to check deprecated values on a list of strings. Updated in my last commit.

Reply:
nice, this wasn't that important of a case to fix I think. thanks for doing this